src/rng/backends/mklcpu/mcg59.cpp

```diff
@@ -0,0 +1,579 @@
+/*******************************************************************************
+* Copyright 2021 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing,
+* software distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions
+* and limitations under the License.
+*
+*
+* SPDX-License-Identifier: Apache-2.0
+*******************************************************************************/
+
+#include <iostream>
+#include <CL/sycl.hpp>
+
+#include "mkl_vsl.h"
+
+#include "oneapi/mkl/exceptions.hpp"
+#include "oneapi/mkl/rng/detail/engine_impl.hpp"
+#include "oneapi/mkl/rng/detail/mklcpu/onemkl_rng_mklcpu.hpp"
+
+#include "cpu_common.hpp"
+
+namespace oneapi {
+namespace mkl {
+namespace rng {
+namespace mklcpu {
+
+using namespace cl;
+
+class mcg59_impl : public oneapi::mkl::rng::detail::engine_impl {
+public:
+    mcg59_impl(cl::sycl::queue queue, std::uint64_t seed)
+            : oneapi::mkl::rng::detail::engine_impl(queue) {
+        vslNewStream(&stream_, VSL_BRNG_MCG59, seed);
+        state_size_ = vslGetStreamSize(stream_);
+    }
+
+    mcg59_impl(const mcg59_impl* other) : oneapi::mkl::rng::detail::engine_impl(*other) {
+        vslCopyStream(&stream_, other->stream_);
+        state_size_ = vslGetStreamSize(stream_);
+    }
+
+    // Buffers APIs
+
+    virtual void generate(const uniform<float, uniform_method::standard>& distr, std::int64_t n,
+                          cl::sycl::buffer<float, 1>& r) override {
+        sycl::buffer<char, 1> stream_buf(static_cast<char*>(stream_), state_size_);
+        queue_.submit([&](sycl::handler& cgh) {
+            auto acc_stream = stream_buf.get_access<sycl::access::mode::read_write>(cgh);
+            auto acc_r = r.get_access<sycl::access::mode::read_write>(cgh);

```

Comment:
Can we use SYCL 2020 accessor interfaces here (maybe under the macro if the code should work with SYCL 1.2.1)?

Reply:
done